server/src/main/java/io/crate/planner/operators/HashJoin.java

```diff
@@ -99,21 +99,34 @@ public ExecutionPlan build(DependencyCarrier executor,
 
         SubQueryAndParamBinder paramBinder = new SubQueryAndParamBinder(params, subQueryResults);
         var hashSymbols = HashJoinConditionSymbolsExtractor.extract(joinCondition);
-        // First extract the symbols that belong to rhs
+        /* It is important here to process the hashSymbols in order as there values are used for building the
+         *  hash codes. For example:
+         *
+         *      join-condition:     t1.a = t2.c AND t1.b = t2.d
+         *      left hashSymbols:   [t1.a, t1.b]
+         *      right hashSymbols:  [t2.c, t2.d]
+         *
+         *      with rows:          left ->     [1, 3]
+         *                          right ->    [1, 3]
+         *
+         * if the order is not guaranteed, one side may use [3, 1] for hash code generation which yields different results
+         *
+         */
         var rhsHashSymbols = new ArrayList<Symbol>();
-        for (var relationName : rhs.getRelationNames()) {
-            var symbols = hashSymbols.remove(relationName);
-            if (symbols != null) {
-                for (var symbol : symbols) {
-                    rhsHashSymbols.add(paramBinder.apply(symbol));
-                }
-            }
-        }
-        // All leftover extracted symbols belong to the lhs
         var lhsHashSymbols = new ArrayList<Symbol>();
-        for (var symbols : hashSymbols.values()) {
+        var rightRelationNames = rhs.getRelationNames();
+        for (var entry : hashSymbols.entrySet()) {
+            var relationName = entry.getKey();
+            var symbols = entry.getValue();
+            if (symbols == null) {
+                continue;
+            }
+            var symbolConsumer = lhsHashSymbols;
```

Comment:
I would prefer to not use a swapping variable in a loop if it can be avoided.

Reply:
why? the var is reset in each iteration.

Reply:
I think this is easier to read:

```
        var rhsHashSymbols = new ArrayList<Symbol>();
        var lhsHashSymbols = new ArrayList<Symbol>();
        var rightRelationNames = rhs.getRelationNames();
        for (var entry : hashSymbols.entrySet()) {
            var relationName = entry.getKey();
            var symbols = entry.getValue();
            if (symbols == null) {
                continue;
            }
            if (rightRelationNames.contains(relationName)) {
                for (Symbol symbol : symbols) {
                    rhsHashSymbols.add(paramBinder.apply(symbol));
                }
            } else {
                for (Symbol symbol : symbols) {
                    lhsHashSymbols.add(paramBinder.apply(symbol));
                }
            }
        }
```

Reply:
However, its a personal preference and minor.

Reply:
Actually, to make the code more resilient I would probably go for this version, to make sure the hash symbols on each side really belong there.

```
        var rhsHashSymbols = new ArrayList<Symbol>();
        var lhsHashSymbols = new ArrayList<Symbol>();
        var rightRelationNames = rhs.getRelationNames();
        var leftRelationNames = lhs.getRelationNames();
        for (var entry : hashSymbols.entrySet()) {
            var relationName = entry.getKey();
            var symbols = entry.getValue();
            if (symbols == null) {
                continue;
            }
            if (rightRelationNames.contains(relationName)) {
                for (Symbol symbol : symbols) {
                    rhsHashSymbols.add(paramBinder.apply(symbol));
                }
            if (leftRelationNames.contains(relationName)) {
                for (Symbol symbol : symbols) {
                    lhsHashSymbols.add(paramBinder.apply(symbol));
                }
            }
        }
```